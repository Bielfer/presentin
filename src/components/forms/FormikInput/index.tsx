/* eslint react-hooks/exhaustive-deps:off */
import {
  ChangeEvent,
  FC,
  InputHTMLAttributes,
  useEffect,
  useState,
} from 'react';
import { useField } from 'formik';
import Input from '@/components/core/Input';

interface Props extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  name: string;
  password?: boolean;
  formatter?: string;
  hint?: string;
}

const FormikInput: FC<Props> = ({
  label,
  name,
  password,
  formatter,
  className,
  hint,
  ...props
}) => {
  const [{ value }, { error }, { setValue }] = useField(name);
  const [formattedValue, setFormattedValue] = useState('');

  const formatValue = (valueParameter: string) => {
    if (!formatter) return valueParameter;

    const fieldValue = valueParameter.split('');
    const notUnderscoreIndexes = [];

    for (let i = 0; i < formatter.length; i += 1) {
      if (formatter[i] === '_') continue;

      notUnderscoreIndexes.push(i);
    }

    for (let i = 0; i < notUnderscoreIndexes.length; i += 1) {
      const toInsertCharacterIndex = notUnderscoreIndexes[i];

      if (toInsertCharacterIndex > fieldValue.length - 1) break;

      fieldValue.splice(
        toInsertCharacterIndex,
        0,
        formatter[toInsertCharacterIndex]
      );
    }

    return fieldValue.join('');
  };

  const removeFormatterCharacters = (valueParameter: string) => {
    if (!formatter) return valueParameter;

    const formatterCharacters: { [key: string]: boolean } = {};

    for (let i = 0; i < formatter.length; i += 1) {
      const formatterCharacter = formatter[i];

      if (formatterCharacter === '_') continue;

      formatterCharacters[formatterCharacter] = true;
    }

    const arrayFormatterCharacters = Object.keys(formatterCharacters);

    return valueParameter
      .split('')
      .filter((character) => !arrayFormatterCharacters.includes(character))
      .join('');
  };

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (!formatter) {
      setValue(e.target.value);
      return;
    }

    const valueWithoutFormatter = removeFormatterCharacters(e.target.value);

    setValue(valueWithoutFormatter);
    setFormattedValue(formatValue(valueWithoutFormatter));
  };

  useEffect(() => {
    if (!formatter) return;

    setFormattedValue(formatValue(value));
  }, []);

  return (
    <Input
      value={formatter ? formattedValue : value}
      onChange={handleChange}
      width="100%"
      password={password}
      label={label}
      className={className}
      error={error}
      hint={hint}
      {...props}
    />
  );
};

export default FormikInput;
