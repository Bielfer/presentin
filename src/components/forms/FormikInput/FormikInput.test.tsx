import { render, screen, waitFor } from '@testing-library/react';
import { Form, Formik } from 'formik';
import * as Yup from 'yup';
import userEvent from '@testing-library/user-event';
import FormikInput from '.';

describe('FormikInput', () => {
  const handleSubmit = jest.fn();

  const validationSchema = Yup.object({
    test: Yup.string().required('required'),
  });

  const Input = (
    <Formik
      initialValues={{ test: '' }}
      validationSchema={validationSchema}
      onSubmit={handleSubmit}
    >
      <FormikInput label="test" name="test" />
    </Formik>
  );

  const InputWithButton = (
    <Formik
      onSubmit={handleSubmit}
      initialValues={{ test: '' }}
      validationSchema={validationSchema}
    >
      {({ submitForm }) => (
        <Form>
          <FormikInput label="test" name="test" />
          <button
            type="submit"
            onClick={() => {
              submitForm();
            }}
          >
            test
          </button>
        </Form>
      )}
    </Formik>
  );

  const InputWithFormatter = (
    <Formik
      initialValues={{ test: '' }}
      validationSchema={validationSchema}
      onSubmit={handleSubmit}
    >
      <FormikInput label="test" name="test" formatter="___.___.___-__" />
    </Formik>
  );

  it('matches snapshot', () => {
    const { asFragment } = render(Input);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the input', () => {
    render(Input);

    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('displays error message, submit after correction', async () => {
    render(InputWithButton);

    await waitFor(() => userEvent.click(screen.getByRole('button')));

    expect(await screen.findByText('required')).toBeInTheDocument();

    await waitFor(() => userEvent.type(screen.getByRole('textbox'), 't'));
    await waitFor(() => userEvent.click(screen.getByRole('button')));

    expect(handleSubmit).toHaveBeenCalled();
  });

  it('should be able to format the text', async () => {
    render(InputWithFormatter);

    const InputElement = screen.getByRole('textbox');

    await waitFor(() => userEvent.type(InputElement, '99999999999'));

    expect(InputElement).toHaveValue('999.999.999-99');
  });
});
