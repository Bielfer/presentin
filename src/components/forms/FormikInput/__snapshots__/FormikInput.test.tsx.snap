// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormikInput matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-between"
    >
      <label
        class="flex items-center text-sm font-medium text-gray-700"
      >
        test
      </label>
    </div>
    <div
      class="mt-1 relative rounded-md shadow-sm"
    >
      <input
        class="shadow-sm focus:ring-blue-500 focus:border-blue-500 block w-full sm:text-sm border-gray-300 rounded-md"
        type="text"
        value=""
        width="100%"
      />
    </div>
  </div>
</DocumentFragment>
`;
